File: src/lib/wb/media.ts
import { wbFetch, type WbCtx } from "./client";

/**
 * Attach media by raw bytes — one file per request. No public hosting needed,
 * so this is the default for a self-hosted generator.
 * Headers: X-Nm-Id, X-Photo-Number (1-based). multipart field name: uploadfile.
 */
export async function uploadMediaBytes(
  ctx: WbCtx,
  nmID: number,
  photoNumber: number,
  buffer: Buffer,
  filename = "photo.jpg",
  contentType = "image/jpeg"
): Promise<void> {
  const form = new FormData();
  const blob = new Blob([new Uint8Array(buffer)], { type: contentType });
  form.append("uploadfile", blob, filename);

  await wbFetch({
    token: ctx.token,
    sandbox: ctx.sandbox,
    method: "POST",
    path: "/content/v3/media/file",
    raw: form,
    headers: {
      "X-Nm-Id": String(nmID),
      "X-Photo-Number": String(photoNumber),
    },
    timeoutMs: 60_000,
  });
}

/**
 * Attach media by public URLs. Field is `nmId` (lowercase d). The whole set is
 * REPLACED, so pass all desired URLs together.
 */
export async function saveMediaByUrls(
  ctx: WbCtx,
  nmID: number,
  urls: string[]
): Promise<void> {
  await wbFetch({
    token: ctx.token,
    sandbox: ctx.sandbox,
    method: "POST",
    path: "/content/v3/media/save",
    body: { nmId: nmID, data: urls },
    timeoutMs: 60_000,
  });
}

File: src-tauri/src/wb/media.rs
#![allow(dead_code)]
//! Media upload by raw bytes (ported from src/lib/wb/media.ts).
//! multipart field: uploadfile; headers: X-Nm-Id, X-Photo-Number (1-based).

use crate::state::AppState;
use crate::wb::client::{wb_fetch, MultipartData, WbCtx, WbReq};
use anyhow::Result;

pub async fn upload_media_bytes(
    state: &AppState,
    ctx: &WbCtx,
    nm_id: i64,
    photo_number: i64,
    bytes: Vec<u8>,
    filename: &str,
) -> Result<()> {
    wb_fetch(
        state,
        ctx,
        WbReq::post("/content/v3/media/file")
            .form(MultipartData {
                field: "uploadfile".into(),
                filename: filename.to_string(),
                content_type: "image/jpeg".into(),
                bytes,
            })
            .header("X-Nm-Id", &nm_id.to_string())
            .header("X-Photo-Number", &photo_number.to_string())
            .timeout(60_000),
    )
    .await?;
    Ok(())
}

/// Upload a product VIDEO to a card. Same `media/file` endpoint as photos, but a
/// video content-type puts it in WB's independent video lane (X-Photo-Number=1
/// does NOT clobber the cover photo); WB transcodes to HLS server-side. Larger
/// timeout — a dubbed mp4 is tens of MB. (Verified on the WB sandbox.)
pub async fn upload_video_bytes(
    state: &AppState,
    ctx: &WbCtx,
    nm_id: i64,
    bytes: Vec<u8>,
    filename: &str,
) -> Result<()> {
    wb_fetch(
        state,
        ctx,
        WbReq::post("/content/v3/media/file")
            .form(MultipartData {
                field: "uploadfile".into(),
                filename: filename.to_string(),
                content_type: "video/mp4".into(),
                bytes,
            })
            .header("X-Nm-Id", &nm_id.to_string())
            .header("X-Photo-Number", "1")
            .timeout(300_000),
    )
    .await?;
    Ok(())
}
